[pytorch] Make rref fetch calls async.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31086 [pytorch] Make rref fetch calls async.**

This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

Commit:
[pytorch] Make rref fetch calls async.

This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -176,19 +176,45 @@ template class UserRRef<py::object>;
 
+template <typename T>
+bool OwnerRRef<T>::hasValue() const {
+  std::unique_lock<std::mutex> lock(mutex_);
```
`std::lock_guard` is supposedly cheaper if you don't have to unlock.

Never measured this or saw any proof though.

jjlilley replied:
ah right, I guess std::lock_guard can avoid a conditional in the destructor. I'll switch.

pietern commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -176,19 +176,45 @@ template class UserRRef<py::object>;
+    }
+  } else {
+    lock.unlock();
+    valueCV_.notify_all();
```
Why not stick to a single branch? Swapping an empty vector is probably only 2 mov's.

jjlilley replied:
Sure, I'll do that - it looks cleaner, and the branching might well cost as much as swapping a few pointers.

pietern reviewed (commented):
LGTM!

I'll leave it to others to approve since there may be dragons hiding that I don't know about.

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -302,6 +307,7 @@ class OwnerRRef final : public RRef {
   c10::optional<T> value_;
   mutable std::mutex mutex_;
   mutable std::condition_variable valueCV_;
+  std::vector<std::shared_ptr<FutureMessage>> futures_;
```
As all the futures will be ready when the value is set, I wonder if it would be sufficient to use a single future with a list of callbacks?

jjlilley replied:
Oh, right... this is where mapping to folly::Future implicitly in my head (where this wouldn't be possible) is counter-productive. :/ I'll replace with a single FutureMessage.

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -176,19 +176,43 @@ template class UserRRef<py::object>;
+  lock.unlock();
   valueCV_.notify_all();
+  if (future.get()) {
+    future->markCompleted();
```
The same future can be mark completed twice in the following scenario? Is this intentional?

1. call `setValue`, so that `value_.has_value() == true`
2. call `getFuture`, a new `future_` will be created and mark completed.
3. call `setValue` again, `future_` will be swapped to `future` and call `markCompleted()` the second time. 

This would fail under current `FutureMessage` implementation. If this is not intentional, let's only mark complete when `future.complete() == false`.

jjlilley replied:
Thanks for bringing this up - I didn't consider the case where setValue() might be called twice.
I'll guard against this.

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

Commit:
rebase on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Thanks!

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Make rref fetch calls async."


This change leverages the new future response framework so that server
threads don't block until setValue is called. Particulurly, we add a
getFuture() method to OwnerRRef so that we get a future that is satisfied
once setValue is called.

Differential Revision: [D18925272](https://our.internmc.facebook.com/intern/diff/D18925272/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -176,19 +176,43 @@ template class UserRRef<py::object>;
+  future.swap(future_);
+  lock.unlock();
   valueCV_.notify_all();
+  if (future.get() && !future->completed()) {
```
@jjlilley 

If a thread calls `ownerRRef->setValue(v)` first, It swaps out a `nullptr` and not `markCompleted(..)` on the new future it put back in.
Later, another thread call `ownerRRef->getFuture()->addCallback()`, this callback will never be executed. We have similar discussion in https://github.com/pytorch/pytorch/pull/32608.

Should we only have one future (no swapping) and initialize it on construction.
